### PR TITLE
[CAPI] CAPI impl refactor

### PR DESCRIPTION
**Pull request by cloudhan**
1. Better options string building
2. avoid potential `new` `delete`

### CODE REVIEW

File: onnxruntime/core/session/provider_bridge_ort.cc [hunk 1]
```diff
@@ -1706,28 +1706,26 @@ ORT_API_STATUS_IMPL(OrtApis::UpdateTensorRTProviderOptions,
   API_IMPL_END
 }
 
+static std::string BuildOptionsString(const onnxruntime::ProviderOptions::iterator& begin,
+                                      const onnxruntime::ProviderOptions::iterator& end) {
+  std::ostringstream options;
+  auto it = begin;
+  if (it != end) {
+    options << it->first << "=" << it->second;
+    it++;
```
Comment (yuslepukhin) [inline onnxruntime/core/session/provider_bridge_ort.cc:1715]: >it++; [](http://example.com/codeflow?start=4&length=5)

Consider prefix ++ here and below

**Comment by yuslepukhin** [comment]
> #endif

Would you consider fixing these things here or in a separate PR? Do a smart ptr, release at the end and remove all these warning suppressions? Thx!

---
Refers to: onnxruntime/core/session/provider_bridge_ort.cc:1786 in 29b4279. [](commit_id = 29b4279c2f76f7c1fd71ed756aff8a94e894e1d7, deletion_comment = False)

**Comment by yuslepukhin** [comment]
>   delete ptr;

Ditto

---
Refers to: onnxruntime/core/session/provider_bridge_ort.cc:1863 in 29b4279. [](commit_id = 29b4279c2f76f7c1fd71ed756aff8a94e894e1d7, deletion_comment = False)

**Comment by yuslepukhin** [comment]
>   *out = new OrtCANNProviderOptions();

Ditto

---
Refers to: onnxruntime/core/session/provider_bridge_ort.cc:1890 in 29b4279. [](commit_id = 29b4279c2f76f7c1fd71ed756aff8a94e894e1d7, deletion_comment = False)

**Comment by yuslepukhin** [comment]
>   delete ptr;

Ditto

---
Refers to: onnxruntime/core/session/provider_bridge_ort.cc:1955 in 29b4279. [](commit_id = 29b4279c2f76f7c1fd71ed756aff8a94e894e1d7, deletion_comment = False)

**Comment by yuslepukhin** [comment]
>   *out = new OrtDnnlProviderOptions();

Ditto

---
Refers to: onnxruntime/core/session/provider_bridge_ort.cc:1978 in 29b4279. [](commit_id = 29b4279c2f76f7c1fd71ed756aff8a94e894e1d7, deletion_comment = False)

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Review by pranavsharma — APPROVED** [review]

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Pull request merged**